broadinstitute/single_cell_portal_core
# Convey limit on number of plots for spatial studies (SCP-5628)

This mitigates problems with rendering too many plots, which most affects spatial transcriptomics studies.

### Overview
Previously, internal web browser constraints caused problems when many scatter plots were rendered for spatial studies.  SCP uses WebGL via Plotly.js to quickly draw scatter plots with many points (i.e. cells).  Sometimes when this occurs, the browser console will throw a warning: "Too many active WebGL contexts. Oldest context will be lost."  But the warning is invisible to users, hard to interpret, and does not reliably appear even in the web browser's DevTools console.  So spatial rendering in the context of high plot-count could silently fail, and simply would not display scatter plots for new groups

Now, SCP shows a brief, actionable warning message when this issue is likely to occur.  This doesn't solve the problem, as that would take substantially more effort, but it does greatly decrease its impact.

### Video
#### Problem
Plots don't display when e.g. > 7 spatial groups are selected, and that limitation is not communicated.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/c5fb223d-09e3-4ffb-bb3e-8b152a78431f

#### Solution
Communicate potential issue with plots when many spatial groups are selected, and how to fix it.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/bdc0af00-da3c-4d9a-8ab0-00f6bc15f359

### Next steps
Future work could consider using non-WebGL Plotly.js contexts, or only rendering plots in viewport, among other more robust solutions.

### Test
New automated tests verify behavior.  If you'd like, you can manually test this like so:
1.  Populate updated synthetic data: `echo "SyntheticStudyPopulator.populate('chicken_raw_counts')" | bundle exec rails console -e development`
2. Go to new "Chicken raw counts" study
3. Select 6 spatial groups
4. Confirm last scatter plot name matches name of last selected spatial group
5. Select 7th spatial group
6. Confirm warning message appears
4. Confirm last scatter plot name does not match name of last selected spatial group
5. Deselect a spatial group, so 7 are selected
6. Confirm warning disappears
7. Select a spatial group, so 8 are selected
8. Confirm warning appears
9. Click "x" to dismiss warning
10. Confirm warning disappears
11. Search for gene "RAD51"
12. Confirm warning appears
13. Remove 4 spatial groups, so 3 are selected
14. Confirm warning disappears

### Further details
This satisfies SCP-5628.  The [2024-06-18 SCP demo video](https://drive.google.com/file/d/1ZjedGN8O-dFpsCI7aD23X7U-rOFawlfF/view?t=19) gives more context.
